waiter/src/waiter/auth/oidc.clj

```diff
@@ -0,0 +1,211 @@
+              (log/info throwable "error in retrieving access token")
+              (utils/exception->response
+                (ex-info (.getMessage throwable)
+                         (-> (ex-data throwable)
+                           (utils/assoc-if-absent :log-level :info)
+                           (utils/assoc-if-absent :status http-401-unauthorized))
+                         throwable)
+                request))))))))
+
+(defn trigger-authorize-redirect
+  "Triggers a 302 temporary redirect response to the authorize endpoint."
+  [jwt-auth-server password {:keys [query-string uri] :as request} response]
+  (let [request-host (utils/request->host request)
+        request-scheme (utils/request->scheme request)
+        code-verifier (create-code-verifier)
+        state-data {:redirect-uri (str (name request-scheme) "://" request-host uri
+                                       (when query-string (str "?" query-string)))}
+        state-code (create-state-code password state-data)
+        authorize-uri (jwt/retrieve-authorize-url
+                        jwt-auth-server request oidc-callback-uri code-verifier state-code)]
+    (-> response
+      (assoc :status http-302-moved-temporarily)
+      (update :headers assoc
+              "cache-control" "no-store"
+              "content-security-policy" content-security-policy-value
+              "location" authorize-uri)
+      (cookie-support/add-encoded-cookie
+        password oidc-challenge-cookie code-verifier challenge-cookie-duration-secs))))
+
+(defn make-oidc-auth-response-updater
+  "Returns a response updater that rewrites 401 waiter responses to 302 redirects."
+  [jwt-auth-server password request]
+  (fn update-oidc-auth-response [{:keys [status] :as response}]
+    (if (and (= status http-401-unauthorized)
+             (utils/waiter-generated-response? response))
+      ;; issue 302 redirect
+      (trigger-authorize-redirect jwt-auth-server password request response)
+      ;; non-401 response, avoid authentication challenge
+      response)))
+
+(defn supports-redirect?
+  "Returns true when:
+   - either the accept-redirect-auth request header is present in the request
+   - or the request is deemed to have come from a browser."
+  [oidc-authority request]
+  (or (hu/browser-request? request)
+      (and (= "yes" (get-in request [:headers accept-redirect-header-name]))
+           (let [accept-redirect-auth (get-in request [:headers accept-redirect-auth-header-name])]
+             (or (str/blank? accept-redirect-auth)
+                 (= "." accept-redirect-auth)
+                 (some #(= oidc-authority %) (str/split accept-redirect-auth #" ")))))))
+
+(defn wrap-auth-handler
+  "Wraps the request handler with a handler to trigger JWT access token authentication."
+  [{:keys [allow-oidc-auth-api? allow-oidc-auth-services? jwt-auth-server oidc-authorize-uri password]} request-handler]
+  (let [oidc-authority (-> oidc-authorize-uri (URI.) (.getAuthority))]
```

Comment:
Yes it can, but then we need to pass extra information (either new key or a sub-map for oidc-authorize-uri) in the authenticator map above.
This is computed once while initializing the auth handler.